public/app/plugins/datasource/prometheus/services/ReferenceSrv.ts

```diff
@@ -0,0 +1,109 @@
+import { clone, each, keyBy } from 'lodash';
+
+export interface QueryWithReference {
+  expr: string;
+  refId: string;
+  refCount?: number;
+}
+
+interface ReferenceSrvProps {
+  initialQueries: QueryWithReference[];
+}
+
+type RefId = string;
+
+export class ReferenceSrv {
+  interpolate() {}
+
+  queries: Map<RefId, QueryWithReference> = new Map();
+
+  getQuery(query: QueryWithReference): QueryWithReference | undefined {
+    return this.queries.get(query.refId);
+  }
+
+  getQueries(): QueryWithReference[] {
+    return Array.from(this.queries).map((mappedQuery) => mappedQuery[1]);
+  }
+
+  setQuery(query: QueryWithReference): void {
+    this.queries.set(query.refId, query);
+  }
+
+  constructor(props: ReferenceSrvProps) {
+    if (props.initialQueries) {
+      props.initialQueries.forEach((query) => {
+        this.queries.set(query.refId, query);
+      });
+    }
+  }
+
+  /**
+   *
+   * @param staleTarget
+   */
+  interpolatePrometheusReferences(staleTarget: QueryWithReference): QueryWithReference {
```

Comment:
still need to save the index, and length of each interpolation, and associate that with the stored query so we can calculate the correct position of the _n_ reference interpolations that happen in each query